### PR TITLE
Add recursive type checking and fallback tests for casting array with unsupported element types to string

### LINKED ISSUE #4446: [BUG] GpuCast crashs when casting from Array with unsupportable child type
**Describe the bug**
Currently, GpuCast will crash with cuDF exception when we try to cast from ArrayType which holds an inconvertable child type towards the cast target.

**Steps/Code to reproduce bug**
Cast Array of Map to String with GPU enabled

**Expected behavior**
In my opinion, compared with adding a P.S. note in the user documentation, a better approach is to verify child types of Array in `CastExprMeta.recursiveTagExprForGpuCheck`, falling back the Cast expression to CPU if there exists non-replaceable type pairs. It will prevent the crash of  GpuCast from the misbehaviors of users.

### PR DESCRIPTION
**Pull request by HaoYang670**
Signed-off-by: remzi <13716567376yh@gmail.com>
close #4446 

1. Add recursive type checking for casting array to string.
2. Add `MapType` into unsupported element types in integration tests
3. Replace `Xfail` by `fallback` when casting unsupported types to string in integration tests

### CODE REVIEW

**Comment by HaoYang670** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala [hunk 1]
```diff
@@ -124,6 +124,9 @@ final class CastExprMeta[INPUT <: CastBase](
           case (fromChild, toChild) =>
             recursiveTagExprForGpuCheck(fromChild.dataType, toChild.dataType, depth + 1)
         }
+      case (ArrayType(element_type, _), StringType) =>
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:127]: nit `element_type` does not follow the convention. Could we make it `elementType`?

**Review by revans2 — APPROVED** [review]
Great catch/fix. Just a minor nit

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala (hunk 1 above)
Comment (HaoYang670) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:127]: Thank you, I will fix it

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 2]
```diff
@@ -1270,10 +1270,11 @@ class CastChecks extends ExprChecks {
   val calendarChecks: TypeSig = none
   val sparkCalendarSig: TypeSig = CALENDAR + STRING
 
-  val arrayChecks: TypeSig = STRING + ARRAY.nested(commonCudfTypes + DECIMAL_128_FULL + NULL +
+  val arrayChecks: TypeSig = psNote(TypeEnum.STRING, "the array's children must also support " +
+    "being cast to string") + ARRAY.nested(commonCudfTypes + DECIMAL_128_FULL + NULL +
       ARRAY + BINARY + STRUCT + MAP) +
-      psNote(TypeEnum.ARRAY, "The array's child type must also support being cast to " +
-          "the desired child type")
+      psNote(TypeEnum.ARRAY, "The array's children must also support being cast to the " +
```
[the remark this answers is not in the document]
Comment (HaoYang670) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:1276]: Thank you, I will fix it

**Comment by HaoYang670** [comment]
build

**Comment by HaoYang670** [comment]
build

**Review by revans2 — APPROVED** [review]

**Pull request merged**